### PR TITLE
Auto-installing requested SDK versions.

### PR DESCRIPTION
**Pull request by associahedron**
Implements A1 of #687, and fixes a bug where requesting an SDK version that is not installed would cause an error (so you couldn't even install the missing version manually).

### CODE REVIEW

File: daml-assistant/src/DAML/Assistant/Env.hs [hunk 1]
```diff
@@ -157,6 +160,12 @@ getSdk damlPath projectPathM =
                     config <- readConfig path
                     fromRightM throwIO (parseVersion config)
 
+        sdkPathFallback :: DamlPath -> Maybe SdkVersion -> IO (Maybe SdkPath)
+        sdkPathFallback damlPath sdkVersionM = do
```
Comment (neil-da) [inline daml-assistant/src/DAML/Assistant/Env.hs]: I find this block pretty confusing. I'd just pattern match on sdkVersionM and simplify it a lot.

File: daml-assistant/src/DAML/Assistant/Env.hs [hunk 2]
```diff
@@ -181,3 +190,41 @@ getDispatchEnv Env{..} = do
            , (sdkVersionEnvVar, maybe "" versionToString envSdkVersion)
            , (damlAssistantEnvVar, unwrapDamlAssistantPath envDamlAssistantPath)
            ] ++ filter ((`notElem` damlEnvVars) . fst) originalEnv
+
+
+-- | Auto-installs requested version if it is missing and updates daml-assistant
+-- if it is the latest stable version.
```
Comment (neil-da) [inline daml-assistant/src/DAML/Assistant/Env.hs]: Why not just if it is newer than the one currently running? I see no requirement to always go to the latest stable. Being always 1 behind would be just fine.

File: daml-assistant/src/DAML/Assistant/Env.hs [hunk 3]
```diff
@@ -181,3 +190,41 @@ getDispatchEnv Env{..} = do
...
+autoInstall :: DamlPath -> Maybe SdkVersion -> IO (Maybe SdkPath)
+autoInstall damlPath sdkVersionM = do
+    damlConfigE <- try $ readDamlConfig damlPath
+    let doAutoInstallME = queryDamlConfig ["auto-install"] =<< damlConfigE
+        doAutoInstallM  = either (const (Just True)) id doAutoInstallME
```
Comment (neil-da) [inline daml-assistant/src/DAML/Assistant/Env.hs]: `either (const True) (fromMaybe True) doAutoInstallME` seems simpler than doing it in two steps.

File: daml-assistant/src/DAML/Assistant/Env.hs [hunk 4]
```diff
@@ -181,3 +190,41 @@ getDispatchEnv Env{..} = do
...
+        doAutoInstall   = fromMaybe True doAutoInstallM
+    whenMaybe (doAutoInstall && isJust sdkVersionM) $ do
+        let sdkVersion = fromJust sdkVersionM
+            sdkPath = defaultSdkPath damlPath sdkVersion
+
+        unlessM (doesDirectoryExist (unwrapSdkPath sdkPath)) $ do
+            -- sdk is missing, so let's install it!
+            -- first, determine if it is the latest stable version
+            latestVersionE :: Either AssistantError SdkVersion
+                <- try getLatestVersion
```
Comment (neil-da) [inline daml-assistant/src/DAML/Assistant/Env.hs]: I think this should just be if its newer than what is currently installed

File: daml-assistant/src/DAML/Assistant/Install.hs [hunk 5]
```diff
@@ -57,6 +59,7 @@ data InstallEnv = InstallEnv
     , targetVersionM :: Maybe SdkVersion
     , damlPath :: DamlPath
     , projectPathM :: Maybe ProjectPath
+    , out :: Handle
```
Comment (neil-da) [inline daml-assistant/src/DAML/Assistant/Install.hs]: Why not `output :: (String -> IO ())`? Restricting it to a handle is still pretty restrictive.

File: daml-assistant/src/DAML/Assistant/Env.hs [hunk 6]
```diff
@@ -139,7 +141,8 @@ getSdk damlPath projectPathM =
 
         sdkPath <- firstJustM id
             [ fmap SdkPath <$> lookupEnv sdkPathEnvVar
-            , pure (defaultSdkPath damlPath <$> sdkVersion)
+            , autoInstall damlPath sdkVersion
```
Comment (cocreature) [inline daml-assistant/src/DAML/Assistant/Env.hs]: It looks like `autoInstall` does roughly the same check as `sdkPathFallback` at the beginning if `auto-install` is enabled and if it isn’t we use `sdkPathFallback` anyway. Can we move `sdkPathFallback` before `autoInstall` and remove the check from `autoInstall`?

Comment (associahedron) [reply daml-assistant/src/DAML/Assistant/Env.hs]: Oh, that's true! :+1:

File: daml-assistant/src/DAML/Assistant/Env.hs (hunk 1 above)
Comment (associahedron) [reply daml-assistant/src/DAML/Assistant/Env.hs]: :+1:

File: daml-assistant/src/DAML/Assistant/Env.hs (hunk 2 above)
Comment (associahedron) [reply daml-assistant/src/DAML/Assistant/Env.hs]: That does make more sense here.

File: daml-assistant/src/DAML/Assistant/Env.hs (hunk 3 above)
Comment (associahedron) [reply daml-assistant/src/DAML/Assistant/Env.hs]: managed to make it even simpler :-)

File: daml-assistant/src/DAML/Assistant/Install.hs (hunk 5 above)
Comment (associahedron) [reply daml-assistant/src/DAML/Assistant/Install.hs]: :+1:

File: daml-assistant/src/DAML/Assistant/Env.hs (hunk 4 above)
Comment (associahedron) [reply daml-assistant/src/DAML/Assistant/Env.hs]: Done!

**Pull request merged**